src/DsaLocal.cc

```diff
@@ -533,6 +533,62 @@ void IntraBlockBuilder::visitLoadInst(LoadInst &LI) {
   }
 }
 
+/// OldVal := *Ptr
+/// if OldVal == Cmp then
+///    *Ptr := New
+/// return OldVal
+void IntraBlockBuilder::visitAtomicCmpXchgInst(AtomicCmpXchgInst &I) {
+  using namespace sea_dsa;
+  
+  if (!m_graph.hasCell(*I.getPointerOperand()->stripPointerCasts())) {
+    return;
+  }
+  Value *Ptr = I.getPointerOperand();
+  Value *Cmp = I.getCompareOperand();
+  Value *New = I.getNewValOperand();
+  
+  Cell PtrC = valueCell(*Ptr);
+  assert(!PtrC.isNull());
+    
+  PtrC.setModified();
+  PtrC.setRead();
+  PtrC.growSize(0, I.getType());
+  PtrC.addAccessedType(0, I.getType());
+
+  if (!isSkip(I)) {
```

Comment:
Isn't `I`'s type a strutcture type `{New->getType(), i1}`?

```
  %13 = cmpxchg weak i64* %8, i64 %11, i64 %12 monotonic monotonic
  %14 = extractvalue { i64, i1 } %13, 0
  %15 = extractvalue { i64, i1 } %13, 1
```

This seems to skip over all the time, no?

Reply:
Fixed